signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,766 @@
+        state_machine
+            .process_message(sender1, nonce_request)
+            .unwrap();
+        assert_allowed_msg_type!(state_machine, SignatureShareResponse);
+        assert_state!(state_machine, State::Signing);
+        assert_message_counts!(state_machine,
+            NonceRequest => total: 1, pending: 0;
+            NonceResponse => total: 2, pending: 0;
+        );
+    }
+
+    #[test]
+    fn test_nonce_phase_with_in_order_messages() {
+        let mut setup = TestSetup::setup(2);
+        let sender1 = setup.sender(0);
+        let sender2 = setup.sender(1);
+        let mut state_machine = setup.state_machine;
+        let mut signer1 = setup.signers.pop().unwrap();
+        let mut signer2 = setup.signers.pop().unwrap();
+
+        assert_eq!(state_machine.signer_count(), 2);
+
+        let nonce_request = nonce_request(1, 1, 1);
+
+        // Process the nonce request with signer 1 and 2 to get their nonce
+        // responses.
+        let nonce_response1 = signer1
+            .process(&nonce_request)
+            .expect("signer1 should be able to process message")
+            .single();
+        assert!(matches!(nonce_response1, Message::NonceResponse(_)));
+        let nonce_response2 = signer2
+            .process(&nonce_request)
+            .expect("signer2 should be able to process message")
+            .single();
+        assert!(matches!(nonce_response2, Message::NonceResponse(_)));
+
+        // The state machine should be able to process the nonce request.
+        state_machine
+            .process_message(sender1, nonce_request)
+            .expect("should be able to process message");
+
+        assert!(state_machine.message_counts(NonceRequest, 1, 0));
+        assert_state!(state_machine, State::Signing);
+        assert!(matches!(
+            state_machine.coordinator.state,
+            wsts::state_machine::coordinator::State::NonceGather(_)
+        ));
+
+        // Process nonce responses from signer 1 and signer 2.
+        state_machine
+            .process_message(sender1, nonce_response1.clone())
+            .expect("should be able to process message");
+        assert_state!(state_machine, State::Signing);
+        assert_allowed_msg_type!(state_machine, NonceResponse);
+        state_machine
+            .process_message(sender2, nonce_response2.clone())
+            .expect("should be able to process message");
+        assert_state!(state_machine, State::Signing);
+
+        // We should have processed all the nonce responses, so we should now be
+        // able to process signature share requests.
+        assert_allowed_msg_type!(state_machine, SignatureShareResponse);
+    }
+
+    #[test]
+    fn test_dkg_verification_state_machine() {
+        let mut setup = TestSetup::setup(2);
+        let sender1 = setup.sender(0);
+        let sender2 = setup.sender(1);
+        let mut state_machine = setup.state_machine;
+        let mut signer1 = setup.signers.pop().unwrap();
```

Comment:
Calling `pop` gives you the signers in reverse order, which was the cause of the issue we saw on Friday.  Since this is test code, why not just use array access here?

Reply:
Ah, good to know. Changed it up a bit in [75c78bc](https://github.com/stacks-network/sbtc/pull/1324/commits/75c78bc3c46868836857bc3750671ccbec4c83c3)